perf/sawtooth_perf/src/batch_submit.rs

```diff
@@ -153,19 +166,343 @@ pub fn http_submitter(target: String, rate: u64, receiver: Arc<Mutex<mpsc::Recei
+        println!("{}", self);
+        println!("Batches/s {} Committed/s {} Invalid/s {}", self.sent_count.load(Ordering::Relaxed) as f64 / update,
+                                                             self.committed_count.load(Ordering::Relaxed) as f64 / update,
+                                                             self.invalid_count.load(Ordering::Relaxed) as f64 / update);
+        self.sent_count.store(0, Ordering::Relaxed);
+        self.committed_count.store(0, Ordering::Relaxed);
+        self.invalid_count.store(0, Ordering::Relaxed);
+    }
+}
+
+
+impl fmt::Display for HTTPRequestCounter {
+
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Since last log, Sent: {0}, Committed {1}/{0}, Invalid {2}/{0} Queue Full {3}/{0}",
+               self.sent_count.load(Ordering::Relaxed),
+               self.committed_count.load(Ordering::Relaxed),
+               self.invalid_count.load(Ordering::Relaxed),
+               self.queue_full_count.load(Ordering::Relaxed))
+    }
+}
+
+/// Run a continuous load of the BatchLists that are generated by BatchListIter.
+pub fn run_workload(batch_list_iter: &mut BatchListIter, time_to_wait: u32, update_time: u32, target: String)
+    -> Result<(), WorkloadError>
+{
+    let mut core = Core::new().unwrap();
+    let handle = core.handle();
+    let handle_clone = handle.clone();
+
+    let mut batch_url = target.clone();
+    batch_url.push_str("/batches");
+
+    let mut last_log = time::Instant::now();
+
+    let counter = Arc::new(HTTPRequestCounter::new());
+    let counter_l = counter.clone();
+    let counter_c = counter.clone();
+    let counter_i = counter.clone();
+    let counter_q = counter.clone();
+    let counter_s = counter.clone();
+
+    let stream : Interval = Interval::new(time::Duration::new(0, time_to_wait), &handle).unwrap();
+    let stream_work = stream
+        .map(move |_| {
+            let log_time = time::Instant::now() - last_log;
+            if log_time.as_secs() as u32 > update_time {
+                counter_l.log(log_time.as_secs(), log_time.subsec_nanos());
+                last_log = time::Instant::now();
+            }
+            match batch_list_iter.next_batch_list()? {
+                Some(batch_list) => Ok(batch_list),
+                None => return Err(WorkloadError::NoBatchError),
+            }
+        }).map_err(|err| WorkloadError::from(err))
+        .map(|batch_list| {
+            let bytes = batch_list?.write_to_bytes()?;
+            let mut req = Request::new(Method::Post, Uri::from_str(batch_url.as_ref())?);
+            let content_len = bytes.len() as u64;
+            req.set_body(bytes);
+            req.headers_mut().set(ContentType::octet_stream());
+            req.headers_mut().set(ContentLength(content_len));
+            Ok(req)
+        }).and_then(|req: Result<Request, WorkloadError>| {
+            let counter_cc = counter_c.clone();
```

Comment:
This seems like a lot of extra cloning

Reply:
There are also a lot of scopes, with each scope passing it off to another closure. These examples in the docs seem to show calling clone() on an Arc being how to share it. https://doc.rust-lang.org/std/sync/struct.Arc.html#examples